torchao/prototype/smoothquant/example.py

```diff
@@ -4,185 +4,263 @@
-            print(f"Saving quantized model to {model_save_path}")
-            t0 = time.time()
-            torch.save(model, model_save_path)
-            print(f"Time to save quantized model: {time.time() - t0:.02f} seconds")
-    if compile:
-        model = torch.compile(model, dynamic=True)
-
-    return benchmark(model, tokenizer, sequence_length, tasks=["PPL"], device=device)
+    model = (
+        AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16)
+        .eval()
+        .to(device)
+    )
+    print(f"Time to load model: {time.time() - t0:.02f} seconds")
 
+    # Step 1: Prepare - insert observers
+    print("running SmoothQuant prepare and calibrate")
+    t0 = time.time()
+    quant_config = SmoothQuantConfig(
+        base_config=Int8DynamicActivationInt8WeightConfig(),
+        step=SmoothQuantStep.PREPARE,
+        alpha=alpha,
+    )
+    quantize_(model, quant_config)
 
-if __name__ == "__main__":
+    # Step 2: Calibration
+    calibration_data = get_calib_dataset(
+        tokenizer=tokenizer, n_samples=calibration_limit, block_size=max_seq_length
+    )
+    for batch in calibration_data:
+        model(batch.to(device))
+        batch.to("cpu")
+
+    print(f"time for prepare and calibration: {time.time() - t0:.02f} seconds")
+
+    # Step 3: Convert to quantized model
+    print("running SmoothQuant convert")
+    t0 = time.time()
+    quant_config.step = SmoothQuantStep.CONVERT
+    quantize_(model, quant_config)
+    print(f"time for convert: {time.time() - t0:.02f} seconds")
+
+    # Set up config for loading
+    quant_config.step = SmoothQuantStep.PREPARE_FOR_LOADING
+    model.config.quantization_config = TorchAoConfig(quant_config)
+
+    if model_save_path is not None:
+        print(f"Saving model to {model_save_path}")
+        torch.save(model, model_save_path)
+
+    if model_save_hf_hub_path is not None:
+        print("pushing model to hub:", model_save_hf_hub_path)
+        model.push_to_hub(model_save_hf_hub_path, safe_serialization=False)
+        tokenizer.push_to_hub(model_save_hf_hub_path)
+
+    print("Benchmarking SmoothQuant model...")
+    return benchmark(model, tokenizer, max_seq_length, tasks=tasks, device=device)
+
+
+def compare_models(
+    model_id: str,
+    alpha: float,
+    tasks: list[str],
+    max_seq_length: int,
+    calibration_limit: int,
+    device: str,
+    model_save_path: str,
+    model_save_hf_hub_path: str,
+):
+    """Compare perplexity and speed for behchmarking SmoothQuant"""
+
+    # Case 1: Base model without quantization
+    print("Benchmarking base model...")
+    torch.manual_seed(34)
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    model = (
+        AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16)
+        .eval()
+        .to(device)
+    )
+    base_results = benchmark(
+        model, tokenizer, max_seq_length, tasks=tasks, device=device
+    )
+
+    # Case 2: W4A8-dynamic without SmoothQuant
```

Comment:
nit: comments and print strings need to be updated from W4A8 to A8W8

although it might be more useful that we should a A8W4 example (in the future)